src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -3283,30 +3819,117 @@ internal override EnumerableSorter<TElement> GetEnumerableSorter(EnumerableSorte
+        {
+            TKey newKey = KeySelector(element);
+            int cmp = Descending ? Comparer.Compare(LastKey, newKey) : Comparer.Compare(newKey, LastKey);
+            if (cacheLower == cmp < 0) LastKey = newKey;
+            return cmp;
+        }
+        internal override void SetElement(TElement element)
+        {
+            LastKey = KeySelector(element);
+        }
+    }
 
+    internal sealed class CachingComparerWithChild<TElement, TKey> : CachingComparer<TElement, TKey>
+    {
+        private readonly CachingComparer<TElement> _child;
+        public CachingComparerWithChild(Func<TElement, TKey> keySelector, IComparer<TKey> comparer, bool descending, CachingComparer<TElement> child)
+            : base(keySelector, comparer, descending)
+        {
+            _child = child;
+        }
+        internal override int Compare(TElement element, bool cacheLower)
+        {
+            TKey newKey = KeySelector(element);
+            int cmp = Descending ? Comparer.Compare(LastKey, newKey) : Comparer.Compare(newKey, LastKey);
+            if (cmp == 0) return _child.Compare(element, cacheLower);
+            if (cacheLower == cmp < 0)
+            {
+                LastKey = newKey;
+                _child.SetElement(element);
+            }
+            return cmp;
+        }
+        internal override void SetElement(TElement element)
+        {
+            base.SetElement(element);
+            _child.SetElement(element);
+        }
+    }
+
+    internal abstract class EnumerableSorter<TElement>
+    {
         internal abstract int CompareKeys(int index1, int index2);
 
-        internal int[] Sort(TElement[] elements, int count)
+        internal abstract void ComputeKeys(TElement[] elements, int count);
+
+        private int[] ComputeMap(TElement[] elements, int count)
         {
             ComputeKeys(elements, count);
             int[] map = new int[count];
             for (int i = 0; i < count; i++) map[i] = i;
+            return map;
+        }
+
+        internal int[] Sort(TElement[] elements, int count)
+        {
+            int[] map = ComputeMap(elements, count);
             QuickSort(map, 0, count - 1);
             return map;
         }
 
+        internal int[] Sort(TElement[] elements, int count, int minIdx, int maxIdx)
+        {
+            int[] map = ComputeMap(elements, count);
+            PartialQuickSort(map, 0, count - 1, minIdx, maxIdx);
+            return map;
+        }
+
+        internal TElement ElementAt(TElement[] elements, int count, int idx)
+        {
+            return elements[QuickSelect(ComputeMap(elements, count), count - 1, idx)];
+        }
+
         private void QuickSort(int[] map, int left, int right)
         {
             do
             {
                 int i = left;
                 int j = right;
-                int x = map[i + ((j - i) >> 1)];
+                int x = j - i < 8 ? map[i + ((j - i) >> 1)] : MedianOfThree(map, i, j);
```

Comment:
I introduced MedianOfThree for the other two partitioning sort/select methods, but it may as well be used here too. It helps avoid quicksort's worse cases.